src/Microsoft.DotNet.Build.Tasks/ValidateProjectDependencyVersions.cs

```diff
@@ -0,0 +1,127 @@
+                            item.ItemSpec);
+                    }
+                }
+                else if (!string.IsNullOrWhiteSpace(_expectedPrerelease))
+                {
+                    log.LogError(
+                        "Both ExpectedVersion and ExpectedPrerelease metadata found on item {0}, but only one permitted",
+                        item.ItemSpec);
+                }
+            }
+
+            public void Validate(string packageId, string version, TaskLoggingHelper log, string dependencyMessage)
+            {
+                var dependencyVersionRange = VersionRange.Parse(version);
+                NuGetVersion dependencyVersion = dependencyVersionRange.MinVersion;
+
+                if (_idPattern.IsMatch(packageId))
+                {
+                    if (!string.IsNullOrWhiteSpace(_expectedVersion) && _expectedVersion != version)
+                    {
+                        log.LogError(
+                            "Dependency validation error: package version '{0}', but expected '{1}' for packages matching '{2}' for {3}",
+                            version,
+                            _expectedVersion,
+                            _idPattern,
+                            dependencyMessage);
+                    }
+                    if (!string.IsNullOrWhiteSpace(_expectedPrerelease) &&
+                        dependencyVersion.IsPrerelease &&
+                        _expectedPrerelease != dependencyVersion.Release)
+                    {
+                        log.LogError(
+                            "Dependency validation error: package prerelease '{0}', but expected '{1}' for packages matching '{2}' for {3}",
+                            dependencyVersion.Release,
+                            _expectedPrerelease,
+                            _idPattern,
+                            dependencyMessage);
+                    }
+                }
+            }
+        }
+
+        [Required]
+        public ITaskItem ProjectJson { get; set; }
+
+        public bool ProhibitFloatingDependencies { get; set; }
+
+        public ITaskItem[] ValidationPatterns { get; set; }
+
+        public override bool Execute()
+        {
+            var patterns = Enumerable.Empty<ValidationPattern>();
+
+            if (ValidationPatterns != null)
+            {
+                patterns = ValidationPatterns
+                    .Select(item => new ValidationPattern(item, Log))
+                    .ToArray();
+            }
+
+            using (TextReader projectFileReader = File.OpenText(ProjectJson.ItemSpec))
+            {
+                var projectJsonReader = new JsonTextReader(projectFileReader);
+
+                var serializer = new JsonSerializer();
+                JObject projectRoot = serializer.Deserialize<JObject>(projectJsonReader);
+
+                foreach (var package in projectRoot["dependencies"].OfType<JProperty>())
```

Comment:
Fixed to look at all descendants to find properties with the name "dependencies" and look through those.